megabus/src/main/java/com/bazaarvoice/megabus/service/KafkaStreamsService.java

```diff
@@ -65,14 +67,20 @@ public KafkaStreamsService(String serviceName,
             _streamsConfiguration.put(SslConfigs.SSL_KEYSTORE_PASSWORD_CONFIG, sslConfiguration.getKeyStorePassword());
             _streamsConfiguration.put(SslConfigs.SSL_KEY_PASSWORD_CONFIG, sslConfiguration.getKeyPassword());
         }
+
+        _streamsExceptionCounter = metricRegistry.counter("bv.emodb.megabus.kafka-streams-exception.count");
     }
 
     protected abstract Topology topology();
 
     @Override
     protected final void doStart() {
         _streams = new KafkaStreams(topology(), _streamsConfiguration);
-        _streams.setUncaughtExceptionHandler((thread, throwable) -> _uncaughtException.compareAndSet(null, throwable));
+        _streams.setUncaughtExceptionHandler((thread, throwable) -> {
+            _uncaughtException.compareAndSet(null, throwable);
+            _streamsExceptionCounter.inc();
+            notifyFailed(_uncaughtException.get());
```

Comment:
Will this work? It seems to me if you call `notifyFailed` directly, the `_streams.close()` will never be called.

Reply:
This should work because `notifyFailed` will cause `ResilientService` to invoke `doStop`, which closes Streams. Previously, `onChange` was only invoked after either the global stream thread died or all of the StreamThreads died, which was the problem we were having with things not restarting even as StreamThreads died. Additionally, calling `Streams#close` from within `onChange` can result in a deadlock because `close`  changes the state.

Reply:
@jonaf That is why i used to call close with a 1 millisecond timeout, as it prevented the deadlock. As far as I can tell, `ResilientService` does not call `stop` in the event of a failure, as I believe that is an invalid operation.